### PR TITLE
Feat: edit  mappings

### PR DESCRIPTION
**Pull request by nguyenthinhhn**
## Related Tickets

- [#TicketID](https://dev.sun-asterisk.com/issues/xxx)

## What's this PR do ?

- [x] update mapping

## Performance

- [x] Resolved n + 1 query
- [x] Run explain query already
- [x] Time run rake task : 1000 ms

## Checklist

- [x] It was tested in local success?
- [x] Fill link PR into ticket and the opposite
- [x] Note purpose/reason, solution, scope of influence into ticket

## ENV note

```
```

## Notes
*(Other notes)*
Screenshot from 2020-01-22 09-40-30

### CODE REVIEW

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 1]
```diff
@@ -39,24 +40,39 @@ public function create(Webhook $webhook)
...
         try {
-            $mapping = $this->mappingRepository->create($attributes);
+            $keys = $request->keys;
+            $values = $request->values;
+            if ($keys && $values) {
+                for ($i = 0; $i < count($keys); $i++) {
+                    $attributes['webhook_id'] = $webhookId;
+                    $attributes['key'] = $keys[$i];
+                    $attributes['value'] = $values[$i];
+                    $mapping = Mapping::where('webhook_id', $webhookId)->updateOrCreate(['key' => $attributes['key']], $attributes);
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 57</br>Line exceeds 120 characters; contains 132 characters

File: app/Http/Controllers/MappingController.php [hunk 2]
```diff
@@ -39,24 +40,39 @@ public function create(Webhook $webhook)
...
-                ->with('messageSuccess', [
-                    'status' => 'Create success',
-                    'message' => 'This mapping successfully created',
-                ]);
-        } catch (QueryException $e) {
-            return back()->with('messageFail', [
-                'status' => 'Create failed',
+            return response()->json([
+                'error' => false,
+                'webhook_id' => $webhook->id
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Arrays.ArrayDeclaration.NoComma</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 69</br>Each line in an array declaration must end in a comma

File: app/Http/Controllers/MappingController.php [hunk 3]
```diff
@@ -80,24 +97,40 @@ public function edit(Webhook $webhook, Mapping $mapping)
...
-            $mapping = $this->mappingRepository->update($mapping->id, $attributes);
+            $mappings = $webhook->mappings()->delete();
+            $keys = $request->keys;
+            $values = $request->values;
+            if ($keys && $values) {
+                for ($i = 0; $i < count($keys); $i++) {
+                    $attributes['webhook_id'] = $webhookId;
+                    $attributes['key'] = $keys[$i];
+                    $attributes['value'] = $values[$i];
+                    $mapping = Mapping::where('webhook_id', $webhookId)->updateOrCreate(['key' => $attributes['key']], $attributes);
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 115</br>Line exceeds 120 characters; contains 132 characters

File: app/Http/Controllers/MappingController.php [hunk 4]
```diff
@@ -80,24 +97,40 @@ public function edit(Webhook $webhook, Mapping $mapping)
...
-                ->with('messageSuccess', [
-                    'status' => 'Update success',
-                    'message' => 'This mapping successfully updated',
-                ]);
-        } catch (QueryException $e) {
-            return back()->with('messageFail', [
-                'status' => 'Update failed',
+            return response()->json([
+                'error' => false,
+                'webhook_id' => $webhook->id
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Arrays.ArrayDeclaration.NoComma</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 127</br>Each line in an array declaration must end in a comma

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>failure</td><td></td></tr></table>

File: app/Http/Controllers/MappingController.php [hunk 5]
```diff
@@ -39,24 +40,41 @@ public function create(Webhook $webhook)
...
-            $mapping = $this->mappingRepository->create($attributes);
+            $keys = $request->keys;
+            $values = $request->values;
+            if ($keys && $values) {
+                for ($i = 0; $i < count($keys); $i++) {
+                    $attributes['webhook_id'] = $webhookId;
+                    $attributes['key'] = $keys[$i];
+                    $attributes['value'] = $values[$i];
+                    $mapping = Mapping::where('webhook_id', $webhookId)->updateOrCreate([
+                        'key' => $attributes['key']
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Arrays.ArrayDeclaration.NoComma</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 58</br>Each line in an array declaration must end in a comma

File: app/Http/Controllers/MappingController.php [hunk 6]
```diff
@@ -80,24 +99,42 @@ public function edit(Webhook $webhook, Mapping $mapping)
...
+            $mappings = $webhook->mappings()->delete();
+            $keys = $request->keys;
+            $values = $request->values;
+            if ($keys && $values) {
+                for ($i = 0; $i < count($keys); $i++) {
+                    $attributes['webhook_id'] = $webhookId;
+                    $attributes['key'] = $keys[$i];
+                    $attributes['value'] = $values[$i];
+                    $mapping = Mapping::where('webhook_id', $webhookId)->updateOrCreate([
+                        'key' => $attributes['key']
```
Comment (framgiaciv3) [inline app/Http/Controllers/MappingController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Arrays.ArrayDeclaration.NoComma</br>**Severity**: error</br>**File**: app/Http/Controllers/MappingController.php</br>**Line**: 118</br>Each line in an array declaration must end in a comma

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr></table>

**Comment by phucqd-1305** [comment]
OK

**Pull request merged**